c++/CMakeLists.txt

```diff
@@ -134,13 +134,8 @@ if(MSVC)
   # TODO(cleanup): Enable higher warning level in MSVC, but make sure to test
   #   build with that warning level and clean out false positives.
 
-  add_compile_options(/wo4503)
-  # Only warn once on truncated decorated names. The maximum symbol length MSVC
-  # supports is 4k characters, which the parser framework regularly blows. The
-  # compiler likes to print out the entire type that went over the limit along
-  # with this warning, which gets unbearably spammy. That said, we don't want to
-  # just ignore it, so I'm letting it trigger once until we find some places to
-  # inject ParserRefs.
+  add_compile_options(/Zc:__cplusplus)
```

Comment:
Without this flag, does everything that worked before still work? I don't want to introduce a new requirement that downstream users use this flag, but I understand that we need the flag in order to enable some features that were disabled before.

Reply:
Not sure what you are asking about. The change (commit) comes from `v2`.

If you are asking about removing `/wo4503`, I thought it might have been an artefact from the past. MSVC generates quite a few different warnings and I did not see why this one should be treated differently. Leaving it there should be fine too, I just did not see a reason.

For adding `/Zc:__cplusplus` I just went with the `v2` change as well, just tweaked it that it works for C++14 `v1` target. If the code uses the `__cplusplus` define then this option should have been set right from the start so it is consistent with the standard.

Reply:
> Not sure what you are asking about. The change (commit) comes from v2.

Right, but v2 is intentionally breaking compat. The 1.x branch, OTOH, should stay stable enough that people can upgrade patch releases without any fear of breakage. I do understand that we probably _should_ have been setting this option all along, but at this point the priority is not to disrupt anyone, and it's not clear to me if this might do that.

I'd be more comfortable if we can leave these flags alone, or maybe if we could set the flag on one of the two MSVC CI targets but leave it off the other one so that we cover both configurations.

Reply:
Ok, I see what you mean. There are two ways (in MSVC) to detect the C++ standard compiler uses at the compilation:
 * `_MSVC_LANG` (https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170) supported from Visual Studio 2015 Update 3
 * `__cplusplus` macro (https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170) which would give meaningful values from Visual Studio 2017 version 15.7, because for this macro to work the compiler option `/Zc:__cplusplus` must be set as well, Otherwise the macro expands to `199711L` which does not match any C++ standard.

It seems that `capnproto` does not use the former at all and instead uses some combination of detecting MSVC alone and using some guessing.

In few places `capnproto` uses `__cplusplus`, but since it does not define `/Zc:__cplusplus` the outcome on MSVC is probably not what was intended.

I have pulled the patch from `v2` and then conveniently used it for differentiating C++20 mode. Now you are right that if someone does not set this compiler option the code won't compile. The right way would be to review all `__cplusplus` uses and make it robust to MSVC by adding the test for `_MSVC_LANG` as well. This way the code should behave consistently with or without `/Zc: __cplusplus` being set.

After that it should not matter in `capnproto` too and the option could be omitted as well.

Otherwise we may face an opposite dilemma, when someone sets this option in his project, but `capnproto` was built without it so the corresponding tests for `__cplusplus` will resolve to different outcomes between the project and the library.

Reply:
I force pushed the corrected patch, which:
* removed `/Zc:__cplusplus` from compile options and instead mimicked it with `_MSVC_LANG` macro, so regardless whether the user sets the option or not, the detection will be governed by `_MSVC_LANG` (for MSVC). `clang-cl` seems to be setting both defines correctly so it should be fine too and any other toolchain will be treated as before. (This fix should not be needed in `v2` as the proper logic will be ensured by `/Zc:__cplusplus` option.)
* I trimmed `"Can trace through coroutines"` from the `async-coroutine test.c++` for `clang-cl` as well as this does not work and the comment at the test claims that the test depends on `GetFunctorStartAddress` function, which test is also skipped on `clang-cl` (and would not work either). So this is more to keep it on par with MSVC were both tests are skipped too.

There is still the crash in `fiber pool limit` test in `Release` build (both in `MSVC` and `clang-cl`), but I guess this is for another patch.

Reply:
Now, on the second thought, the fix for `/Zc:__cplusplus` might be a safe approach even on `v2` against cases where users would not set this option in their projects. So the header files which use `__cplusplus` logic will not get treated differently in the library build and the users' projects builds.

Reply:
After another thought I think it would make sense to define internal define to cover both `__cplusplus` and `_MSVC_LANG` for the purpose of `kj` lib. Something like:
```c++
#if defined(_MSVC_LANG)
#define KJ_CPP_STD _MSVC_LANG
#else
#define KJ_CPP_STD __cplusplus
#endif
```
possibly in `kj/common.h`?

I pushed a commit which applies the changes described above.